src/backend/vulkan/src/window.rs

```diff
@@ -34,69 +38,124 @@ impl Drop for RawSurface {
-                            }
-                        };
-
-                        let info = vk::XlibSurfaceCreateInfoKHR {
-                            s_type: vk::StructureType::XlibSurfaceCreateInfoKhr,
-                            p_next: ptr::null(),
-                            flags: vk::XlibSurfaceCreateFlagsKHR::empty(),
-                            window: window.get_xlib_window().unwrap() as *const _,
-                            dpy: window.get_xlib_display().unwrap() as *mut _,
-                        };
-
-                        unsafe {
-                            xlib_loader.create_xlib_surface_khr(&info, None).ok()
-                        }
-                    }
-                    #[cfg(windows)]
-                    vk::VK_KHR_WIN32_SURFACE_EXTENSION_NAME => {
-                        use kernel32;
-                        use winit::os::windows::WindowExt;
-
-                        let win32_loader = ext::Win32Surface::new(entry, &self.raw.0)
-                            .expect("Unable to load win32 surface functions");
-
-                        unsafe {
-                            let info = vk::Win32SurfaceCreateInfoKHR {
-                                s_type: vk::StructureType::Win32SurfaceCreateInfoKhr,
-                                p_next: ptr::null(),
-                                flags: vk::Win32SurfaceCreateFlagsKHR::empty(),
-                                hinstance: kernel32::GetModuleHandleW(ptr::null()) as *mut _,
-                                hwnd: window.get_hwnd() as *mut _,
-                            };
-
-                            win32_loader.create_win32_surface_khr(&info, None).ok()
-                        }
-                    }
-                    // TODO: other platforms
-                    _ => None,
-                }
-            })
-            .find(|x| x.is_some())
-            .expect("Unable to find a surface implementation.")
-            .unwrap();
+        let surface = {
+            use std::mem::transmute;
+
+            let xlib_loader = ext::XlibSurface::new(entry, &self.raw.0)
+                                .expect("XlibSurface failed");
+
+            let info = vk::XlibSurfaceCreateInfoKHR {
+                s_type: vk::StructureType::XlibSurfaceCreateInfoKhr,
+                p_next: ptr::null(),
+                flags: vk::XlibSurfaceCreateFlagsKHR::empty(),
+                window: unsafe { transmute(window) },
+                dpy: display as *mut _,
+            };
+
+            unsafe { xlib_loader.create_xlib_surface_khr(&info, None) }
+                .expect("XlibSurface failed")
+        };
+
+        let (width, height) = unsafe {
+            use x11::xlib::XGetGeometry;
+            use std::mem::{uninitialized, transmute};
+            use std::os::raw::{c_int, c_uint, c_ulong};
+            let mut root_return: c_ulong = uninitialized();
+            let mut x_return: c_int = uninitialized();
```

Comment:
why not `=0`?

Reply:
Variables with `*_return` names above are filled by `XGetGeometry`, so it is reasonable leaving it undefined on initialization IMO. BTW, `XGetGeometry` will fail if window is an invalid XID, it should fire panic.

Reply:
Besides, we should cast `Window` as `c_ulong` to pass to `XGetGeometry ` and `XGetWindowAttributes` (due to `drawable` parameter is defined as `c_ulong` (aka `XID`)

Reply:
I understand that it should work. I don't understand why we would want them to be uninitialized (more typing, more confusion during any debugging session) as opposed to simple zeroes.

Reply:
Because I feel that it's weird to modify the same variables twice...

Reply:
I replaced `uninitialized()` with `0` to avoid confusion. Besides, I should not think too much about optimization since the code path is a cold path, not hot path.